src/plugins/plugin-serializable.ts

```diff
@@ -0,0 +1,55 @@
+import type { Plugin, AnyFn, Ctx } from '../types';
+import isPlainObject from 'lodash.isplainobject';
+
+/**
+ * Check if a value can be serialized (e.g. using `JSON.stringify`).
+ * Adapted from: https://stackoverflow.com/a/30712764/3829557
+ */
+function isSerializable(value: any) {
+  // Primitives are OK.
+  if (
+    value === undefined ||
+    value === null ||
+    typeof value === 'boolean' ||
+    typeof value === 'number' ||
+    typeof value === 'string'
+  ) {
+    return true;
+  }
+
+  // A non-primitive value that is neither a POJO or an array cannot be serialized.
+  if (!isPlainObject(value) && !Array.isArray(value)) {
+    return false;
+  }
+
+  // Recurse entries if the value is an object or array.
+  for (const key in value) {
+    if (!isSerializable(value[key])) return false;
+  }
+
+  return true;
+}
+
+/**
+ * Plugin that checks whether state is serializable, in order to avoid
+ * network serialization bugs.
+ */
+const SerializablePlugin: Plugin = {
+  name: 'plugin-serializable',
+
+  fnWrap: (move: AnyFn) => (G: unknown, ctx: Ctx, ...args: any[]) => {
+    const result = move(G, ctx, ...args);
+    /* istanbul ignore if */
+    if (process.env.NODE_ENV === 'production') {
+      // Do not perform this check in production.
+      return result;
+    }
+    if (!isSerializable(result)) {
+      throw new Error(`Move state is not JSON-serialiazable. See 
+      https://boardgame.io/documentation/#/?id=state for more information.`);
+    }
```

Comment:
This can be simplified like this I think (which also avoids the need for an istanbul ignore comment). I’ve adjusted the error message as the template literal would include the indentation after the new line.

```suggestion
    // Check state in non-production environments.
    if (process.env.NODE_ENV !== 'production' && !isSerializable(result)) {
      throw new Error(
        'Move state is not JSON-serialiazable.\n' +
          'See https://boardgame.io/documentation/#/?id=state for more information.'
      );
    }
```

Reply:
In general, I think the use of guard statements reduces the nesting of the code and makes it easier to read, but in this specific case either way works as the code is quite simple